File: booking-service/src/infrastructure/messaging/EventBridgePublisher.ts
import { EventBridgeClient, PutEventsCommand } from '@aws-sdk/client-eventbridge';
import { EventPublisher } from '../../domain/port/EventPublisher';
import { DomainEvent } from '../../domain/event/DomainEvent';

/**
 * EventBridgePublisher - Adapter for publishing domain events
 * 
 * Implements EventPublisher port using AWS EventBridge.
 * 
 * Hexagonal Architecture:
 * - Domain defines what it needs (EventPublisher interface)
 * - Infrastructure implements how (this class)
 * - Domain doesn't know about EventBridge
 * 
 * Event-Driven Architecture:
 * - Publishes domain events to EventBridge
 * - Other services can subscribe to these events
 * - Enables loose coupling between services
 */
export class EventBridgePublisher implements EventPublisher {
  private readonly client: EventBridgeClient;
  private readonly eventBusName: string;
  private readonly source: string;

  constructor(
    client: EventBridgeClient,
    eventBusName: string = 'event-management-bus',
    source: string = 'booking-service'
  ) {
    this.client = client;
    this.eventBusName = eventBusName;
    this.source = source;
  }

  /**
   * Publishes a domain event to EventBridge
   */
  async publish(event: DomainEvent): Promise<void> {
    const command = new PutEventsCommand({
      Entries: [
        {
          Source: this.source,
          DetailType: event.eventType,
          Detail: JSON.stringify(event),
          EventBusName: this.eventBusName,
          Time: event.occurredAt,
        },
      ],
    });

    const result = await this.client.send(command);

    // Check if event was published successfully
    if (result.FailedEntryCount && result.FailedEntryCount > 0) {
      const error = result.Entries?.[0]?.ErrorMessage || 'Unknown error';
      throw new Error(`Failed to publish event: ${error}`);
    }
  }

  /**
   * Publishes multiple domain events in a batch
   * EventBridge supports up to 10 events per batch
   */
  async publishBatch(events: DomainEvent[]): Promise<void> {
    if (events.length === 0) {
      return;
    }

    // EventBridge limit: 10 events per batch
    const batches = this.chunkArray(events, 10);

    for (const batch of batches) {
      const command = new PutEventsCommand({
        Entries: batch.map(event => ({
          Source: this.source,
          DetailType: event.eventType,
          Detail: JSON.stringify(event),
          EventBusName: this.eventBusName,
          Time: event.occurredAt,
        })),
      });

      const result = await this.client.send(command);

      if (result.FailedEntryCount && result.FailedEntryCount > 0) {
        throw new Error(`Failed to publish ${result.FailedEntryCount} events`);
      }
    }
  }

  /**
   * Helper: Splits array into chunks
   */
  private chunkArray<T>(array: T[], size: number): T[][] {
    const chunks: T[][] = [];
    for (let i = 0; i < array.length; i += size) {
      chunks.push(array.slice(i, i + size));
    }
    return chunks;
  }
}

File: payment-service/src/infrastructure/messaging/eventbridge_publisher.py
"""
EventBridge Publisher - Event publishing adapter

Implements EventPublisher port using AWS EventBridge.
Publishes domain events to event bus for other services.
"""

import os
import json
from typing import List
from datetime import datetime, timezone
import boto3
from botocore.exceptions import ClientError

from ...domain.ports import EventPublisher
from ...domain.events import DomainEvent


class EventBridgePublisher(EventPublisher):
    """EventBridge implementation of EventPublisher"""
    
    def __init__(self, event_bus_name: str = None):
        """
        Initialize publisher
        
        Args:
            event_bus_name: EventBridge bus name (defaults to env var)
        """
        self.event_bus_name = event_bus_name or os.getenv(
            "EVENT_BUS_NAME",
            "event-management-bus"
        )
        
        # Configure EventBridge client for LocalStack
        endpoint_url = os.getenv("AWS_ENDPOINT_URL", "http://localhost:4566")
        region = os.getenv("AWS_DEFAULT_REGION", "us-east-1")
        
        self.client = boto3.client(
            "events",
            endpoint_url=endpoint_url,
            region_name=region,
            aws_access_key_id=os.getenv("AWS_ACCESS_KEY_ID", "test"),
            aws_secret_access_key=os.getenv("AWS_SECRET_ACCESS_KEY", "test")
        )
    
    async def publish(self, event: DomainEvent) -> None:
        """
        Publish a single domain event
        
        Args:
            event: Domain event to publish
            
        Raises:
            Exception: If publish fails
        """
        await self.publish_batch([event])
    
    async def publish_batch(self, events: List[DomainEvent]) -> None:
        """
        Publish multiple domain events
        
        Args:
            events: List of domain events to publish
            
        Raises:
            Exception: If publish fails
        """
        if not events:
            return
        
        try:
            entries = [
                self._create_event_entry(event)
                for event in events
            ]
            
            response = self.client.put_events(Entries=entries)
            
            # Check for failures
            if response.get("FailedEntryCount", 0) > 0:
                failed = [
                    entry for entry in response.get("Entries", [])
                    if "ErrorCode" in entry
                ]
                raise Exception(f"Failed to publish {len(failed)} events: {failed}")
                
        except ClientError as e:
            raise Exception(f"Failed to publish events: {e.response['Error']['Message']}")
    
    def _create_event_entry(self, event: DomainEvent) -> dict:
        """
        Create EventBridge entry from domain event
        
        Args:
            event: Domain event
            
        Returns:
            EventBridge entry dictionary
        """
        return {
            "Time": datetime.now(timezone.utc),
            "Source": "payment-service",
            "DetailType": event.event_type,
            "Detail": json.dumps(event.to_dict()),
            "EventBusName": self.event_bus_name
        }
